packages/shared/src/contexts/PushNotificationContext.tsx

```diff
@@ -0,0 +1,150 @@
+import React, {
+  createContext,
+  ReactElement,
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import { useQuery, useQueryClient } from '@tanstack/react-query';
+import OneSignal from 'react-onesignal';
+import { AnalyticsEvent, NotificationPromptSource } from '../lib/analytics';
+import { disabledRefetch } from '../lib/func';
+import { useAuthContext } from './AuthContext';
+import { generateQueryKey, RequestKey } from '../lib/query';
+import { isTesting } from '../lib/constants';
+import { useAnalyticsContext } from './AnalyticsContext';
+import { SubscriptionCallback } from '../components/notifications/utils';
+import { BootApp } from '../lib/boot';
+
+export interface PushNotificationsContextData {
+  OneSignal: typeof OneSignal;
+  isPushSupported: boolean;
+  isInitialized: boolean;
+  isSubscribed: boolean;
+  isLoading: boolean;
+  shouldOpenPopup: boolean;
+  onSourceChange: (source: string) => void;
+  trackPermissionGranted: (source: NotificationPromptSource) => void;
+}
+
+export const PushNotificationsContext =
+  createContext<PushNotificationsContextData>({
+    isPushSupported: false,
+    isInitialized: true,
+    isSubscribed: false,
+    isLoading: false,
+    shouldOpenPopup: true,
+    trackPermissionGranted: null,
+    onSourceChange: null,
+    OneSignal: null,
+  });
+
+interface PushNotificationContextProviderProps {
+  children: ReactElement;
+  app: BootApp;
+}
+
+export function PushNotificationContextProvider({
+  children,
+  app,
+}: PushNotificationContextProviderProps): ReactElement {
+  const notificationSourceRef = useRef<string>();
+  const onSourceChange = useCallback((source) => {
+    notificationSourceRef.current = source;
+  }, []);
+  const subscriptionCallbackRef = useRef<SubscriptionCallback>();
+  const [isSubscribed, setIsSubscribed] = useState(false);
+  const { user } = useAuthContext();
+  const { trackEvent } = useAnalyticsContext();
+  const isEnabled = !!user && !isTesting;
+  const key = generateQueryKey(RequestKey.OneSignal, user);
+  const client = useQueryClient();
+  const {
+    data: OneSignalCache,
+    isFetched,
+    isLoading,
+  } = useQuery<typeof OneSignal>(
+    key,
+    async () => {
+      const osr = client.getQueryData<typeof OneSignal>(key);
+
+      if (osr) {
+        return osr;
+      }
+
+      await OneSignal.init({
+        appId: process.env.NEXT_PUBLIC_ONESIGNAL_APP_ID,
+        serviceWorkerParam: { scope: '/push/onesignal/' },
+        serviceWorkerPath: '/push/onesignal/OneSignalSDKWorker.js',
+      });
+
+      await OneSignal.login(user.id);
+
+      setIsSubscribed(OneSignal.User.PushSubscription.optedIn);
+
+      OneSignal.User.PushSubscription.addEventListener(
+        'change',
+        ({ current }) => {
+          setIsSubscribed(() => current.optedIn);
+          subscriptionCallbackRef.current?.(current.optedIn);
+        },
+      );
```

Comment:
When do we clean this listener? This could introduce memory leak with listener being added each time queryFn is evaluated.

Reply:
Oh, right. We currently don't clean it up, but the evaluation only happens once and we ensure it by having some early checks..

Reply:
Hmm, I that is a lot of assumptions. Better way would be to do `useEffect` on `OneSignal` value that query returns and then properly subscribe/unsubscribe when you want.

```js
useEffect(() => {
  if (!OneSignal) {
    return undefined;
  }

  const onChange = ({ current }) => setIsSubscribed(() => current.optedIn);

  OneSignal.User.PushSubscription.addEventListener('change', onChange);

  return () => {
    OneSignal.User.PushSubscription.removeEventListener('change', onChange);
  };
}, [OneSignal]);
```

Reply:
Yeah, we can do that.

Reply:
I think it is more than an assumption as it is pretty straightforward. Also OneSignal throws an error whenever you initialize it more than once. From that alone, we understand that the evaluation only happened once.